### ISSUE TITLE
Slows down when updating the 'events' array with many events

### ISSUE THREAD

**Issue by james-interest** [acro5piano/react-native-big-calendar#563]
I've noticed that there is longer rendering time as the number of items in the `events` prop increases. I recorded a video when there's a low and high number of events so show the difference.

From what I understand, day events are render in this part https://github.com/acro5piano/react-native-big-calendar/blob/master/src/components/CalendarBody.tsx#L158-L165

I haven't done any performance measuring to understand where the time is spend to re-render.

Maybe it's React that takes time to render a large list of items on the screen, so maybe there's nothing we can do. Does anyone have any suggestions for something to try or trying another way to render the items? I'm happy to spend time on it but I'm not sure what to try so opening this issue for discussion...

**Video with a high number of events (slows 2 seconds wait time to appear)**

https://user-images.githubusercontent.com/59870708/144372421-9f5ed14f-c68e-48e7-ae9a-d96b6ae298b1.mp4

**Video with a low number of events**

https://user-images.githubusercontent.com/59870708/144372849-4862508d-0282-4aaa-8bf0-0d9983e00dd8.mp4

**Comment by acro5piano** [comment]
Hi @james-interest , thanks for your interest in this lib.

Which Android version are you testing on? If 11, this issue might be related: https://issuetracker.google.com/issues/207695804?pli=1

**Comment by james-interest** [comment]
Hi @acro5piano thanks to you for creating this library 🙏 
I've been noticing the same slowness on the iOS simulator so I don't think it's necessarily related to Android. But it's only noticeable when incrementally adding new items in the `events` array so it's a pretty specific scenario.

The cause might be in the time it takes to render each event on the calendar. I tried doing the same test with a dummy list of squares and the time to render all the square doesn't increase as the total number of squares increases, so that leads me to believe the slow render could be in the library code rather than React performance.

Also, I haven't noticed it on the web, only with RN. I'll try by measuring times in CalendarEvent.tsx: https://github.com/acro5piano/react-native-big-calendar/blob/master/src/components/CalendarEvent.tsx#L43-L66

If my understanding is correct, the `getEventCellPositionStyle` method is called for every event when the `events` array changes, even if only one item was added. Perhaps there's a way to cache the result of `getEventCellPositionStyle` so on subsequent renders, it will use the cached value if there's one. But I have no idea if the time is really spent there so I'll do some testing first by using console.log in different places of the codebase and measuring times.

**Comment by acro5piano** [comment]
Thanks for the investigation. 

How many events are you rendering? A reproduce repository is really helpful!

**Comment by james-interest** [comment]
On the iOS simulator it starts being noticeably slower when there are over 50 events. On Android it would be less maybe 20. I have a scenario where the user can quickly add many blocks of time. It's fine most of the time but when the users adds many events it can get slower, it is a pretty specific edge case...

Here's the repo I set up https://github.com/james-interest/rn-big-calendar-demo

**Comment by successful-fella** [comment]
I have same problem. Tested with both Android and iOS. Changing month, event click and everything slows down very much (takes 4-5s) when event array contains many event.

In my case, I have 98 entries in a single month. Here's the snippet from code -
```
events: [
  {
    "event_id": "2122",
    "start": "2022-05-02 08:30:00",
    "start_timestamp": 1651501800,
    "end": "2022-05-02 10:30:00",
    "title": "Chad Allen",
    "color": "A6D06E"
  } **x 98**
]

<BigCalendar
        events={this.state.events}
        height={600}
        mode={'month'}
        onPressCell={ (date) => {this.showDateEvents(date)} } // Same issue even if I remove this prop
        date={ this.state.calendarDate } // Same issue even if I remove this prop
        onChangeDate={ this.setMonth } // Same issue even if I remove this prop
/>
```

Please can anyone confirm if it works correctly for them on Android/iOS with about 100 events?

Thank you.

**Comment by jordoh** [comment]
After experiencing a slowdown in month view with many events, as described, profiling pointed to a large proportion of time spent in sort, which appears to be due to the order of the `sort` and `filter` in the month view: https://github.com/acro5piano/react-native-big-calendar/blob/f738fc36d67533e3dbdcc9c4b1e2d0dfb5e7a5e8/src/components/CalendarBodyForMonthView.tsx#L175-L190

That is: all events are being sorted for every day rendered, not just the events that fall on that day. Re-ordering these calls so the sort only happens after filtering to the events that fall on the day resulted in a very appreciable speed up. 

Additionally, converting the `sort` and `filter` functions there to use date-fns instead of dayjs and pre-calculating the duration in seconds for each event resulted in a further speed up, to the point where the monthly view is usable with a large number of events.

**Comment by acro5piano** [comment]
Thanks for your investigation! I'm pretty busy for now but would like to try to improve it if I have time.

**Comment by fabrianibrahim** [comment]
any updates to fix this issue?

**Comment by acro5piano** [comment]
Unfortunately no updates for now. @jordoh 's solution sounds viable, so if someone creates a PR for this, I am happy to accept.

> Additionally, converting the sort and filter functions there to use date-fns instead of dayjs and pre-calculating the duration in seconds for each event resulted in a further speed up, to the point where the monthly view is usable with a large number of events.

**Comment by ylahssini** [comment]
Any updates on this issue? because the problem still persisting

**Comment by hdrabs** [comment]
I am having the same issue for the `week` mode, not using month mode.

The workaround I am using is that I am creating a bucket for time range of 1 hour and showing the count on the event cell and when you click on it, it goes to `day` mode. Which is still taking some time to render but better than rendering the events. 

Please if anyone have another solution please let me know

**Comment by arisyo13** [comment]
This issue makes the calendar almost useless. when rendering over 100 events it takes couples seconds to load and same issue when changing the mode. Any update on this? this seems a major issue

**Comment by proparnaampit** [comment]
no update on this? this issue is a major one, need help

**Comment by arisyo13** [comment]
> no update on this? this issue is a major one, need help

Im working on this. expect a PR soon

**Comment by arisyo13** [comment]
> @arisyo13 Please let us know if you find a way to optimize it! 🙏🏼

i opened this pr https://github.com/acro5piano/react-native-big-calendar/pull/1124

**Comment by arisyo13** [comment]
@acro5piano when you time could you check the PR?

**Comment by acro5piano** [comment]
@arisyo13 Thanks, I'll check it shortly

**Comment by acro5piano** [comment]
@arisyo13 's PR is merged and released on v4.16.1. Let me summerize the probelm:

We have the event overlapping feature. The logic is checking every events, then find overlapping event(s) and add margin. Our previous logic was naive and too heavy if there are lots of events. Its time complexity was `O(n^2)`. While `O(n^2)` looks okay for such as UI component, but actually there are comparison two dates (start end end) for each events, which is not silly.

His tuning resolved the time complexity by binary search to find overlapping events. It makes the logic efficient enough to be `O(log n)`. Also, the loop now breaks early because the events are sorted.

Correct me if I'm wrong it.

Let me close this issue. According to his benchmark, rendering 500 events used to took `62 seconds` but now it takes `430 milliseconds` which is really fast.
Big thank you for @arisyo13 again!
